### ISSUE TITLE
utf-8 characters not converted

### ISSUE THREAD

**Issue by mauriciojaramillo** [wkhtmltopdf/wkhtmltopdf#3068]
Dear developers

We are having troubles with some characters conversion in the oficial release, but the alpha release works fine with this chars. Attached you may see the superscript chars (4, 5, ..., 9) on the HTML file (last chars) and they are not converted on PDF result file (pipe chars shown).

Have you a date planed to release this alpha?

Thanks in advance and regards.

PD: nice tool, very useful for our apps.
[issue.zip](https://github.com/wkhtmltopdf/wkhtmltopdf/files/411554/issue.zip)

Mauricio Jaramillo

**Comment by matt-hoskins** [comment]
Hi,

I don't think it's a problem with utf-8 handling, I think it's a font problem. If you look at the fonts embedded in the PDFs the problem pdf has one font, the working one has two.

In your problem PDF I think the font used to render is likely missing characters at those codepoints and no substitution has been performed.

If I run the above file through wkhtmltopdf on one of my servers which has a few fonts available (version 0.12.3 with patched qt under linux) it comes out ok, albeit with the font slightly larger (as no font or font size is specified in the html sample). The resulting PDF has 4 fonts embedded.

![prueba_linux_0 12 3](https://cloud.githubusercontent.com/assets/5127538/17617205/b6fc1dd0-6070-11e6-9d93-4151ccf6998d.png)


I'm guessing you're using the windows version. Perhaps try specifying a specific font family in the html which you know has those unicode codepoints covered and it may well render ok under the official release.

**Comment by mauriciojaramillo** [comment]
Hi dear Matt

 

Thanks a lot for you help, very useful for us. I will let you know our results.

 

Best regards.

 

Mauricio Jaramillo Velez 
_____________________
 <mailto:mauricio@connectividad.net> mauricio@connectividad.net 
Connectividad.net Ltda.
Soluciones Web
 <http://www.connectividad.net/> www.connectividad.net

* Los acentos y caracteres extendidos han sido omitidos voluntariamente en este mensaje para evitar conflictos de visualizacion en los programas de correo electronico de los destinatarios.
* La informacion contenida en este mensaje, incuido todo archivo anexo, es de caracter confidencial y solo debera ser utilizada por el destinatario original. Esta prohibida la divulgacion, difusion y distribucion a terceros sin la previa autorizacion del remitente. Si usted ha recibido este mensaje por error por favor informe al remitente del mismo y eliminelo de su programa de correo electronico. 

 

 <http://www.connectividad.net/flornet/> FlorNet - Gestión Integral de Cultivo -  <http://www.connectividad.net/einventarios/> E-Inventarios - Inventarios Generales y Selectivos

 

 

 

 

De: matt-hoskins [mailto:notifications@github.com] 
Enviado el: viernes, 12 de agosto de 2016 3:40
Para: wkhtmltopdf/wkhtmltopdf
CC: mauriciojaramillo; Author
Asunto: Re: [wkhtmltopdf/wkhtmltopdf] utf-8 characters not converted (#3068)

 

Hi,

I don't think it's a problem with utf-8 handling, I think it's a font problem. If you look at the fonts embedded in the PDFs the problem pdf has one font, the working one has two.

In your problem PDF I think the font used to render is likely missing characters at those codepoints and no substitution has been performed.

If I run the above file through wkhtmltopdf on one of my servers which has a few fonts available (version 0.12.3 with patched qt under linux) it comes out ok, albeit with the font slightly larger (as no font or font size is specified in the html sample). The resulting PDF has 4 fonts embedded.

 <https://cloud.githubusercontent.com/assets/5127538/17617205/b6fc1dd0-6070-11e6-9d93-4151ccf6998d.png> prueba_linux_0 12 3

I'm guessing you're using the windows version. Perhaps try specifying a specific font family in the html which you know has those unicode codepoints covered and it may well render ok under the official release.

**Comment by mauriciojaramillo** [comment]
Hi Matt

Thanks a lot, your way was right. Let me explain.

We need to add some lines on the HEAD / STYLE of the HTML file and run conversion and VOILA, all characters are shown in the PDF file. Below the lines added (omit +++):

+++
<head>
<title> xxxx </title>
<style>
@font-face {
            font-family: "Arial Unicode";
            src: url(.../arialms/ARIALUNI.TTF);
            unicode-range: U+0-10FFFF;
        }

        * {
            font-family: "Arial Unicode";
        }
</style>
+++

In addition we need to copy to the server the file "ARIALUNI.TTF" in a folder "arialms"; this file is located (on Windows) in the path "C:\windows\fonts".

Thanks again dear Matt and Regards
Mauricio Jaramillo

**Comment by matt-hoskins** [comment]
Great to hear you've managed to resolve the problem by setting the font using a style declaration :). 

I've used Arial Unicode MS on a windows server application before (not with wkhtmltopdf though - this was with ActivePDF for rendering Word documents) - we just installed it on the server like other fonts. I think the font's name for use in font-family would be "Arial Unicode MS" if you were to just reference the installed version without needing a src. I guess you're not able to install it as a normal font on your server in your situation though.
